aten/src/ATen/core/jit_type_base.h

```diff
@@ -48,9 +48,78 @@ namespace c10 {
   _(AnyListType)            \
   _(AnyTupleType)           \
   _(AnyClassType)           \
+  _(SymbolicOrConcreteIntType) \
   _(UnionType)              \
   _(DynamicType)
 
+class TORCH_API SymbolicInt {
+public:
+  virtual SymbolicInt* add(SymbolicInt* b) = 0;
+  virtual SymbolicInt* add(int64_t) = 0;
+};
+
+class TORCH_API SymbolicOrConcreteInt {
+
+public:
+
+  SymbolicOrConcreteInt(int64_t d):
+  data_(d) {};
+
+  int64_t data_;
+
+  static const int64_t SYMBOLIC_INT_TAG = 1LL << 63;
+
+  operator int64_t() {
+    TORCH_CHECK(!isSymbolicInt());
+    return data_;
+  }
+
+  bool isSymbolicInt() {
+    return isSymbolicInt(data_);
+  }
+
+  static bool isSymbolicInt(int64_t data) {
+    return data < -1;
+  }
+
+  // TODO: can we make it more C++ cast-style like?
+  static SymbolicInt* unbox(SymbolicOrConcreteInt sci) {
+    TORCH_INTERNAL_ASSERT(sci.isSymbolicInt());
+    return reinterpret_cast<SymbolicInt*>(sci.data_ ^ SYMBOLIC_INT_TAG);
+  }
+
+  static int64_t box(SymbolicInt* s) {
+    return reinterpret_cast<int64_t>(s) ^ SYMBOLIC_INT_TAG;
+  }
+  
+
+  bool operator==(const SymbolicOrConcreteInt& p2)
+  {
+      TORCH_INTERNAL_ASSERT("NYI");
+      return false;
+  }
+
+  SymbolicOrConcreteInt operator+(SymbolicOrConcreteInt sci) {
+    int64_t result = 0;
+    if (this->isSymbolicInt()) {
+      if (sci.isSymbolicInt()) {
+        result = box(unbox(*this)->add(unbox(sci)));
+      }
+      else {
+        result = box(unbox(*this)->add(sci.data_));
+      }
+    } else if (sci.isSymbolicInt()) {
+        result = box(unbox(sci)->add(data_));
+    } else {
+      result = data_ + sci.data_;
+    }
+
+    return SymbolicOrConcreteInt(result);
+  }
+};
```

Comment:
this stuff doesn't need to be in the symint only PR, can just be a straight wrapper